ray_on_golem/server/services/ray.py

```diff
@@ -187,34 +191,59 @@ async def _get_head_node(self) -> Node:
                 if node.tags.get(TAG_RAY_NODE_KIND) == NODE_KIND_HEAD:
                     return node
 
-    def _is_head_node_to_webserver_tunel_running(self) -> bool:
-        return self._head_node_to_webserver_tunel_process is not None
+    def _is_head_node_to_webserver_tunnel_running(self) -> bool:
+        return self._head_node_to_webserver_tunnel_process is not None
 
-    async def _start_head_node_to_webserver_tunel(self) -> None:
-        head_node = await self._get_head_node()
+    async def _start_head_node_to_webserver_tunnel(self) -> None:
+        logger.info("Starting head node to webserver tunnel...")
 
-        self._head_node_to_webserver_tunel_process = await start_ssh_reverse_tunel_process(
-            str(head_node.internal_ip),
-            self._golem_service._ray_on_golem_port,
-            private_key_path=self._ssh_private_key_path,
-            proxy_command=head_node.ssh_proxy_command,
+        head_node = await self._get_head_node()
+        port = self._golem_service._ray_on_golem_port
+
+        self._head_node_to_webserver_tunnel_process = await run_subprocess(
+            "ssh",
+            "-N",
+            "-R",
+            f"*:{port}:127.0.0.1:{port}",
+            "-o",
+            "StrictHostKeyChecking=no",
+            "-o",
+            "UserKnownHostsFile=/dev/null",
+            "-o",
+            f"ProxyCommand={head_node.ssh_proxy_command}",
+            "-i",
+            str(self._ssh_private_key_path),
+            f"{self._ssh_user}@{str(head_node.internal_ip)}",
+        )
+        self._head_node_to_webserver_tunnel_early_exit_task = asyncio.create_task(
+            self._on_head_node_to_webserver_tunnel_early_exit()
         )
 
-        # TODO: Add log when process dies early
+        logger.info("Starting head node to webserver tunnel done")
 
-        logger.info("Reverse tunel from remote head node to local webserver started")
+    async def _on_head_node_to_webserver_tunnel_early_exit(self) -> None:
+        await self._head_node_to_webserver_tunnel_process.communicate()
 
-    async def _stop_head_node_to_webserver_tunel(self) -> None:
-        process = self._head_node_to_webserver_tunel_process
+        logger.warning(f"Head node to webserver tunnel exited prematurely!")
 
-        if process is None:
+    async def _stop_head_node_to_webserver_tunnel(self) -> None:
+        process = self._head_node_to_webserver_tunnel_process
+
+        if process is None or process.returncode is not None:
```

Comment:
Why not ask for forgiveness instead of asking for permission here?
```python
try:
    self._head_node_to_webserver_tunnel_early_exit_task.cancel()
 ...
 ```

Reply:
The only reason is to have explicit logs. Same (or even more) applies to handling of the yagna process.